### PR TITLE
avoid alpn allocation on windows

### PR DESCRIPTION
**Pull request by wfurt**
follow up on #69098. Removes buffer allocations for common case. Pre-computes common sets. 

contributes to #68951

### CODE REVIEW

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs [hunk 1]
```diff
@@ -67,8 +65,45 @@ public static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProto
...
+                }
+                else
+                {
+                    int protocolLenngth = Interop.Sec_Application_Protocols.GetProtocolLength(alpn);
+                    int bufferLength = sizeof(Interop.Sec_Application_Protocols) + protocolLenngth;
+
+                    Span<byte> alpnBuffer;
+                    if (bufferLength <= 64)
+                    {
+                        // workaround for CS8350
```
Comment (jkotas) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs:95]: If you declare the local as `scoped Span<byte> alpnBuffer`, you should not need this workaround.

Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs]: nice. updated.

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs [hunk 2]
```diff
@@ -61,14 +59,40 @@ public static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProto
...
-            InputSecurityBuffers inputBuffers = default;
+            scoped InputSecurityBuffers inputBuffers = default;
             inputBuffers.SetNextBuffer(new InputSecurityBuffer(inputBuffer, SecurityBufferType.SECBUFFER_TOKEN));
             inputBuffers.SetNextBuffer(new InputSecurityBuffer(default, SecurityBufferType.SECBUFFER_EMPTY));
 
             if (context == null && sslAuthenticationOptions.ApplicationProtocols != null && sslAuthenticationOptions.ApplicationProtocols.Count != 0)
             {
-                byte[] alpnBytes = ConvertAlpnProtocolListToByteArray(sslAuthenticationOptions.ApplicationProtocols);
-                inputBuffers.SetNextBuffer(new InputSecurityBuffer(new ReadOnlySpan<byte>(alpnBytes), SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                var alpn = sslAuthenticationOptions.ApplicationProtocols;
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs:68]:
```suggestion
                List<SslApplicationProtocol> alpn = sslAuthenticationOptions.ApplicationProtocols;
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs [hunk 3]
```diff
@@ -61,14 +59,40 @@ public static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProto
...
+                    inputBuffers.SetNextBuffer(new InputSecurityBuffer(s_http1, SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                }
+                else if (alpn.Count == 1 && alpn[0] == SslApplicationProtocol.Http2)
+                {
+                    inputBuffers.SetNextBuffer(new InputSecurityBuffer(s_http2, SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                }
+                else if (alpn.Count == 2 && alpn[0] == SslApplicationProtocol.Http11 && alpn[1] == SslApplicationProtocol.Http2)
+                {
+                    inputBuffers.SetNextBuffer(new InputSecurityBuffer(s_http12, SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs:81]:
```suggestion
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs [hunk 4]
```diff
@@ -98,13 +122,39 @@ public static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProto
...
 
-            InputSecurityBuffers inputBuffers = default;
+            scoped InputSecurityBuffers inputBuffers = default;
             inputBuffers.SetNextBuffer(new InputSecurityBuffer(inputBuffer, SecurityBufferType.SECBUFFER_TOKEN));
             inputBuffers.SetNextBuffer(new InputSecurityBuffer(default, SecurityBufferType.SECBUFFER_EMPTY));
             if (context == null && sslAuthenticationOptions.ApplicationProtocols != null && sslAuthenticationOptions.ApplicationProtocols.Count != 0)
             {
-                byte[] alpnBytes = ConvertAlpnProtocolListToByteArray(sslAuthenticationOptions.ApplicationProtocols);
-                inputBuffers.SetNextBuffer(new InputSecurityBuffer(new ReadOnlySpan<byte>(alpnBytes), SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                var alpn = sslAuthenticationOptions.ApplicationProtocols;
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs:130]:
```suggestion
                List<SslApplicationProtocol> alpn = sslAuthenticationOptions.ApplicationProtocols;
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs [hunk 5]
```diff
@@ -98,13 +122,39 @@ public static byte[] ConvertAlpnProtocolListToByteArray(List<SslApplicationProto
...
+                }
+                else
+                {
+                    int protocolLenngth = Interop.Sec_Application_Protocols.GetProtocolLength(alpn);
+                    int bufferLength = sizeof(Interop.Sec_Application_Protocols) + protocolLenngth;
+                    scoped Span<byte> alpnBuffer = bufferLength <= 64 ? stackalloc byte[bufferLength] : new byte[bufferLength];
+
+                    Interop.Sec_Application_Protocols.SetProtocols(alpnBuffer, alpn, protocolLenngth);
+                    inputBuffers.SetNextBuffer(new InputSecurityBuffer(alpnBuffer, SecurityBufferType.SECBUFFER_APPLICATION_PROTOCOLS));
+                }
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs:157]: Is this if/else chain an exact duplicate of the one earlier in the file?  Can it be extracted into a shared helper?

Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs]: updated into helper function.

**Review by stephentoub — APPROVED** [review]

**Comment by wfurt** [comment]
WASM failures are unrelated.

**Pull request merged**